### DISCUSSION TITLE
Branch coverage is always 0 / namespace grouping (PHPUnit / Clover)

### DISCUSSION THREAD

**Discussion by uuf6429** [danielpalme/ReportGenerator#578] [Q&A]
Title says it all 😄 

Both the HTML report from ReportGenerator and the "bitbucket report" from `LoremFooBar/reportgenerator-coverage-bitbucket-pipe` specify a branch coverage of 0%. I don't know what value to expect, but I'm pretty sure it shouldn't be a 0.

According to PHPUnit this metric should be generated: https://phpunit.readthedocs.io/en/9.5/code-coverage-analysis.html#:~:text=line%20was%20executed.-,Branch%20Coverage,-The%20Branch%20Coverage however, I don't know enough about clover to tell where it should be and if it's missing or in the wrong place.

Given all these uncertainties, I decided to start a discussion instead of report a bug.

----

On a different point, namespace level 1/2 grouping doesn't seem to work well: if I'm not mistaken, the namespace should correspond to a PHP namespace, instead it seems that the namespace is just the class.
For example:
- file name: `src/Traits/MakesHttpResponsesTrait.php`
- namespace (in HTML report): `src/Traits/MakesHttpResponsesTrait.php`
- real PHP namespace: `App\Traits` (and what I was expecting)
- clover xml from PHPUnit:
   ```xml
    <file name="/Users/xx/PhpstormProjects/myapp/src/Traits/MakesHttpResponsesTrait.php">
      <class name="App\Traits\MakesHttpResponsesTrait" namespace="global">
   ```
It feels to me that the class element should look like `<class name="MakesHttpResponsesTrait" namespace="App\Traits">`, shouldn't it? If so, I can do some research on PHPUnit's side and start asking question. 😆

**Comment by danielpalme** [comment]
**Regarding branch coverage:**
Could you please share a sample Clover file generated by PhpUnit?
You can send the file privately by email: reportgenerator@palmmedia.de

Branch coverage is only available if the coverage file contains element of  `type="cond"`.
E.g. `<line falsecount="0" truecount="1" num="15" type="cond"/>`

---

**Regarding namespace levels:**
You are right, the namespaces for PHP are not interpreted perfectly.
That's because other tools that generate Clover reports, put the the namespace into the `name` attribute and add another `path` attribute for the file's path. PhpUnit only uses a `name` attribute.

**Reply by uuf6429** [reply]
**Regarding branch coverage:**

The coverage file does not have any `<line>` elements with a `type="cond"` attribute at all. Strangely, there are `*conditionals` attributes on `metrics` elements that always seem to have a zero value:

![image](https://user-images.githubusercontent.com/230049/209425228-7e641348-06ed-474d-9ca3-7c74154fd4ae.png)

From what I can see, there's no way to reliably extract branch/conditional coverage from PHPUnit's clover file. 😞 

----

**Regarding namespace levels:**

It seems that the "global" namespace value has been implemented like that [since the beginning](https://github.com/sebastianbergmann/php-code-coverage/commit/0c8db36e2bfbb642a89e05d8d7e6a70e6dc7922d#diff-0e105740559ef4553f3ff8c3e5e6dc0db0eb9df4e5b8142a36b5aa958f372b81R106).

I think you could theoretically fix this with minimal risk. Perhaps appending the following (untested) code to `CloverReportPreprocessor::Execute` could work:

```cs
            foreach (var @class in file.Descendants("class").ToArray())
            {
                string className = @class.attribute.Attribute("name").Value;
                int classNamePos = className.LastIndexOf('\\');
                if (classNamePos != -1 && @class.attribute.Attribute("namespace").Value == "global")
                {
                    @class.attribute.Attribute("name").Value = className.Substring(0, classNamePos);
                    @class.attribute.Attribute("namespace").Value = className.Substring(classNamePos + 1);
                }
            }
```

----

On a different point, I think I'll switch to a [different coverage output format](https://phpunit.readthedocs.io/en/9.5/textui.html?highlight=clover#command-line-options) - the lack of a clover standard is an ongoing problem for everyone, so it makes sense to switch to a better option.

**Comment by danielpalme** [comment]
I applied the change in [3e83d9c](https://github.com/danielpalme/ReportGenerator/commit/3e83d9c3842f798b54eda8ff5abb07420619f985).
Next release will contain the improvement.
